File: packages/test/polyglot/python/core_identity.py
from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any

from dbx_tools.core import fnv_hash, to_identifier, to_stable_key


def _decode_value(value: Any) -> Any:
    if isinstance(value, list):
        return [_decode_value(item) for item in value]
    if not isinstance(value, dict):
        return value

    descriptor_type = value.get("$type")
    if descriptor_type == "negativeZero":
        return -0.0
    if descriptor_type == "set":
        return {_decode_value(item) for item in value.get("values", [])}
    if descriptor_type == "cycle":
        cyclic: dict[str, Any] = {}
        cyclic["self"] = cyclic
        return cyclic
    if descriptor_type == "nan":
        return float("nan")
    return {key: _decode_value(item) for key, item in value.items()}


def main() -> None:
    fixture_path = Path(sys.argv[1] if len(sys.argv) > 1 else "fixtures/core-identity.json")
    cases = json.loads(fixture_path.read_text())
    results = []
    for case in cases:
        try:
            if case["operation"] == "fnvHash":
                options = {"length": case["length"]} if "length" in case else {}
                result = fnv_hash(case["value"], **options)
            elif case["operation"] == "toStableKey":
                result = to_stable_key(_decode_value(case["value"]))
            else:
                result = to_identifier(*case["values"])
            results.append({"name": case["name"], "result": result})
        except TypeError as cause:
            results.append({"name": case["name"], "error": type(cause).__name__})
    print(json.dumps(results, separators=(",", ":")))


if __name__ == "__main__":
    main()

File: packages/test/polyglot/src/core-identity.ts
import { readFileSync } from "node:fs";

import { hash, object, string } from "@dbx-tools/shared-core";

interface ExpectedResult {
  result?: string;
  error?: string;
}

interface FnvHashCase {
  name: string;
  operation: "fnvHash";
  value: string;
  length?: number;
  expected: ExpectedResult;
}

interface StableKeyCase {
  name: string;
  operation: "toStableKey";
  value: unknown;
  expected: ExpectedResult;
}

interface IdentifierCase {
  name: string;
  operation: "toIdentifier";
  values: string[];
  expected: ExpectedResult;
}

export type CoreIdentityCase = FnvHashCase | StableKeyCase | IdentifierCase;

export interface CoreIdentityResult extends ExpectedResult {
  name: string;
}

export function readCoreIdentityCases(path: string): CoreIdentityCase[] {
  return JSON.parse(readFileSync(path, "utf8")) as CoreIdentityCase[];
}

export function runTypeScriptCoreIdentityCases(cases: CoreIdentityCase[]): CoreIdentityResult[] {
  return cases.map((testCase) => {
    try {
      let result: string;
      switch (testCase.operation) {
        case "fnvHash":
          result = hash.fnvHashWithOptions(
            testCase.length === undefined ? {} : { length: testCase.length },
            testCase.value,
          );
          break;
        case "toStableKey":
          result = object.toStableKey(decodeValue(testCase.value));
          break;
        case "toIdentifier":
          result = string.toIdentifier(...testCase.values);
          break;
      }
      return { name: testCase.name, result };
    } catch (cause) {
      return {
        name: testCase.name,
        error: cause instanceof Error ? cause.name : typeof cause,
      };
    }
  });
}

function decodeValue(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(decodeValue);
  if (value === null || typeof value !== "object") return value;

  const descriptor = value as { $type?: string; values?: unknown[] };
  switch (descriptor.$type) {
    case "negativeZero":
      return -0;
    case "set":
      return new Set((descriptor.values ?? []).map(decodeValue));
    case "cycle": {
      const cyclic: Record<string, unknown> = {};
      cyclic.self = cyclic;
      return cyclic;
    }
    case "nan":
      return Number.NaN;
    default:
      return Object.fromEntries(
        Object.entries(value).map(([key, item]) => [key, decodeValue(item)]),
      );
  }
}
